### ISSUE TITLE
E1136: False positive when used with typing (mypy) patterns

### ISSUE THREAD

**Issue by ssbarnea** [PyCQA/pylint#3979]
Apparently even the master branch code does not yet play well with a [common pattern](https://stackoverflow.com/a/59313490/99834) used to add type hints. The linked example is not unique, the same pattern can be seen in many places.

While adding ``# pylint: disable=unsubscriptable-object`` can be used as a temporary workaround, it worth creating a bug as the more people will be adopting type hints, the more such problems we will face.

```python
import os
from typing import TYPE_CHECKING, Any, Union

if TYPE_CHECKING:
    BasePathLike = os.PathLike[Any] # <-- that is where pylint identifies E1136
else:
    BasePathLike = os.PathLike

foo : Union[str, BasePathLike] = "bar"
```

### Current behavior
```
example.py:5:19: E1136: Value 'os.PathLike' is unsubscriptable (unsubscriptable-object)
```

### Expected behavior

Pass the linting as the code below is correct.

### pylint --version output
2.6.0 and master, on python 3.9

**Comment by ajeetdsouza** [comment]
This failed for me too:

```python
from typing import Optional
x: Optional[int]
```

**Comment by hippo91** [comment]
@ssbarnea thanks for your report. I can reproduce it.
@ajeetdsouza your issue is different from original one and has already been fixed thanks to #3890

**Comment by cdce8p** [comment]
This seems to be related to #3951
Which will be fixed with the next `astroid` update: PyCQA/astroid#885

**Comment by hippo91** [comment]
@cdce8p thank you! I can confirm the bug is fixed. I can no longer reproduce it with current master branches of `astroid` and `pylint`.

**Comment by alkatar21** [comment]
I have the same message with exactly this code in Python 3.8 is this intentional or a bug?

pylint 2.12.2
astroid 2.9.3
Python 3.8.10

**Comment by Pierre-Sassoulas** [comment]
I can reproduce the issue again with python 3.8.

**Comment by cdce8p** [comment]
The issue was indeed fixed, but only for Python 3.9+.
For 3.8, it's related to #4369, although with a different pattern.

**Comment by Avasam** [comment]
I'm not entirely certain if related, but I'm having the same issue with `Column` type from` the sqlalchemy typestubs
pylint 2.12.2
Python 3.9.6

```py
from __future__ import annotations
from sqlalchemy import Column, DateTime, Integer, String

class Player(BaseModel):
    __tablename__ = "player"

    user_id = db.Column(db.String(8), primary_key=True)
    name = db.Column(db.String(32), nullable=False)
    # The biggest region code I found so far was "us/co/coloradosprings" at 21
    country_code = db.Column(db.String(24))
    score = db.Column(db.Integer, nullable=False)
    score_details = db.Column(db.String())
    last_update = db.Column(db.DateTime())
    rank: Optional[int] = None

    schedules = db.relationship("Schedule", back_populates="owner")

    @overload
    def __init__(  # type: ignore
        self,
        user_id: str | Column[String],
        name: str | Column[String],
        country_code: Optional[str | Column[String]],
        score: int | float | Column[Integer],
        last_update: Optional[str | Column[DateTime]],
        score_details: Optional[str | Column[String]] = ...,
        rank: Optional[int] = ...
    ): ...
    
[...]
```
![image](https://user-images.githubusercontent.com/1350584/153691419-01f83367-726a-48bd-b704-f651768b7b7a.png)

**Comment by cdce8p** [comment]
> I'm not entirely certain if related, but I'm having the same issue with `Column` type from` the sqlalchemy typestubs pylint 2.12.2 Python 3.9.6

@Avasam That shouldn't be an issue anymore. Please make sure to test the latest pylint release. Additionally, stringified annotations should be used (either `"str | Column[String]"` or `from __future__ import annotations`). Then no `unsubscriptable-object` warning is emitted.
